src/subject_reduction.ml

```diff
@@ -0,0 +1,43 @@
+open Terms
+open Print
+open Basics
+
+module Eset = Set.Make(
+  struct
+    type t = (term * term) 
+    let compare = Pervasives.compare
+  end)
+
+exception No_relation
+
+(** [find_relation t] calculates a pair of {!type:term} [A] and {!type:Eset.t}
```

Comment:
No need to rewrite yet another type inference algorithm. Just use Typing.infer_constr like in the current sr.ml file. The difference with the new algorithm is after this point: instead of extracting a substitution from the constraints, we extract a set of rules.